src/content/en/updates/2019/02/hotpath-with-wasm.md

```diff
@@ -0,0 +1,492 @@
+at address 4 and our output image will start directly after the input image
+ends:
+
+    int bpp = 4;
+    int imageSize = inputWidth * inputHeight * bpp;
+    uint32_t* inBuffer = (uint32_t*) 4;
+    uint32_t* outBuffer = (uint32_t*) (imageSize + 4);
+
+    for (int d2 = d2Start; d2 >= 0 && d2 < d2Limit; d2 += d2Advance) {
+      for (int d1 = d1Start; d1 >= 0 && d1 < d1Limit; d1 += d1Advance) {
+        int in_idx = ((d1 * d1Multiplier) + (d2 * d2Multiplier));
+        outBuffer[i] = inBuffer[in_idx];
+        i += 1;
+      }
+    }
+
+Note: The reason we chose to start at address 4 and not 0 is because address 0
+has a special meaning in many languages: It's the dreaded null pointer. While
+technically 0 is a perfectly valid address, many languages exclude 0 as a valid
+value for pointers and either throw an exception or just tumble into undefined
+behavior.
+
+After porting the entire JavaScript function to C, we can compile [the C
+file](https://gist.github.com/surma/0eb306fa9acc8bdf2f58150b2f1e82b4#file-rotate-c)
+with `emcc`:
+
+    $ emcc -O3 -s ALLOW_MEMORY_GROWTH=1 -o c.js rotate.c
+
+As always, emscripten generates a glue code file called `c.js` and a wasm module
+called `c.wasm`. Note that the wasm module gzips to only ~260 Bytes, while the
+glue code is around 3.5KB after gzip. After some fiddling, we were able to ditch
+the glue code and instantiate the WebAssembly modules with the vanilla APIs.
+This is often possible with Emscripten as long as you are not using anything
+from the C standard library.
+
+Note: We are working with the Emscripten team to make the glue code smaller or
+even non-existent at times.
+
+#### Rust
+
+Rust is a new, modern systems programming language. Amongst other things, it
+aims to be useful in the same places that C has been used traditionally, but
+brings modern language concepts and tooling. Rust also supports WebAssembly as a
+first-class citizen and the Rust team has contributed a lot of excellent tooling
+to the WebAssembly ecosystem.
+
+One of these tools is [`wasm-pack`](https://rustwasm.github.io/wasm-pack/),
+which is mostly written by [Ashley Williams](https://twitter.com/ag_dubs) and
+[Nick Fitzgerald](https://twitter.com/fitzgen?lang=en). `wasm-pack` takes your
+code and turns it into a web-friendly npm package that works out-of-the-box with
+bundlers like webpack. `wasm-pack` is an extremely convenient experience, but
+currently only works for Rust. The plan is to add support for other
+WebAssembly-targeting languages in the future.
+
+In Rust, slices are what arrays are in C. And just like in C, we need to create
+slices that use our start addresses. This goes against the memory safety model
+that Rust enforces, so to get our way we have to use the `unsafe` keyword,
+allowing us to write code that doesn't comply with that model.
+
+Note: This is not a best practice. In our experience it is usually worth it to
```

Comment:
is there a best practice example we could give? it's tough to only give one example *and* state that it's not best practice. usually people will ignore that not best practice bit...

Reply:
Fair point! I’ll add a best-practice example to the gist. Sound good?